### PR TITLE
[feat] Add opt-in VLM LoRA training support

### PR DESCRIPTION
**Pull request by FangHeng**
## Description

This PR adds optional PEFT/LoRA support for fine-tuning the StarVLA VLM backbone.

The feature is disabled by default and keeps the existing full fine-tuning, frozen-VLM training, checkpoint loading, and non-LoRA launch behavior unchanged. When enabled, LoRA wraps the configured VLM backbone module and saves PEFT adapters alongside the normal StarVLA checkpoints.

This is intended as a maintained candidate path for memory-constrained adaptation, adapter exchange, and ablation experiments. It is not presented as the default recommendation for best task performance.

## Motivation

Issue [#229](https://github.com/starVLA/starVLA/issues/229) points out a real gap in the current training options:

- freezing the VLM only trains the action side;
- full fine-tuning is expensive and creates large checkpoints;
- the existing adapter path adapts the action head, not the VLM backbone;
- users who want PEFT/LoRA currently need to carry private patches.

Besides its speed and efficiency, LoRA is quite important for methods that inject extra geometric info (like spatial forcing) to avoid corrupting the pre-trained backbone's distribution. We understand the project experience that LoRA fine-tuning may produce poor results, and that users targeting stronger task performance may be better served by the smaller model. This PR keeps that caveat in place. The goal is to make LoRA available as an opt-in, reproducible training mode, not to replace full fine-tuning or the recommended smaller model path.

## Changes

- Added lazy PEFT LoRA utilities for the VLM backbone in `starVLA/model/modules/vlm/lora_utils.py`.
- Added canonical config support under `framework.vlm.lora`, while keeping `framework.qwenvl.lora` as a compatibility fallback.
- Wrapped the configured VLM module before `accelerator.prepare`, so LoRA parameters are visible to optimizer setup.
- Rebuilt optimizer/scheduler setup after PEFT wrapping when LoRA is enabled.
- Added LoRA-aware checkpoint handling:
  - save `vlm_lora_adapter` beside full checkpoints;
  - optional adapter-only save;
  - latest adapter checkpoint discovery;
  - `from_pretrained()` loading for LoRA-shaped checkpoints;
  - legacy `qwen_lora_adapter` fallback.
- Added shared launcher wiring through `examples/common/vlm_lora_args.sh`.
- Wired direct `train_starvla.py` launchers across supported benches to accept `VLM_LORA_*` environment variables.
- Added an optional LoRA-oriented ZeRO-1 DeepSpeed config.
- Added `docs/LoRA.md` with support scope, configuration, launch examples, checkpoint behavior, and smoke checks.

### Support scope

Supported:

- Qwen-series VLM backbones mounted at `qwen_vl_interface.model`.
- Benches that launch `starVLA/training/train_starvla.py`, including Robotwin, LIBERO, SimplerEnv/OXE, DOMINO, Franka, Calvin, Robocasa, Robocasa365, RoboChallenge Table30v2, VLA-Arena, Gemma4, MiniCPM, and NeuralVLA launchers.

Experimental:

- Non-Qwen VLM backbones, if the framework exposes a PEFT-compatible backbone module.
- For non-Qwen models, users can set `framework.vlm.lora.module_path` and use explicit target module names or `target_modules=all-linear`.

Not included:

- WM4A internal world-model LoRA for DiT/T5/VAE. That changes a different part of the stack and should be handled as a separate design decision.

## Testing

- [x] Local training for N steps without errors
- [x] Benchmark evaluation results
- [x] Local LoRA unit/config/checkpoint smoke checks
- [x] Remote NVIDIA LoRA forward/backward/save smoke checks

## Type of Change

- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] New framework / benchmark integration
- [ ] Breaking change
- [ ] Documentation only
- [ ] Refactor (no behavior change)

## Checklist

- [x] No unrelated changes mixed in
- [x] New features have example config in `examples/`
- [x] No secrets, API keys, or large binaries committed
- [x] Files stay isolated — no unnecessary modification to shared modules (`starVLA/dataloader/`, `starVLA/training/`, `starVLA/config/`)
- [x] No modifications to shared files, or justified above

Shared-file justification:

- `starVLA/training/train_starvla.py` is touched because LoRA must be injected after checkpoint/freeze setup and before distributed prepare.
- `starVLA/training/trainer_utils/trainer_tools.py` is touched for LoRA adapter-only checkpoint discovery and resume handling.
- `starVLA/config/deepseeds/` is touched only to add an optional LoRA-oriented ZeRO-1 config.

### CODE REVIEW

**Comment by chongchong789** [comment]
Hi, thanks for adding VLM LoRA support. I would like to clarify the intended checkpoint-saving behavior when jointly training VLM LoRA and the action head.

My planned setup is:

* freeze the original VLM parameters;
* train LoRA adapters on `qwen_vl_interface.model`;
* fully fine-tune `action_model`;
* use the resulting checkpoint for evaluation.

I have not yet completed an actual train-save-reload test, so the following is only my interpretation based on reading the code.

When the configuration is:

```yaml
framework:
  vlm:
    lora:
      enabled: true
      save_adapter_only: true
```

it appears that the normal full-model checkpoint may be skipped.

The relevant saving logic seems to be located in:

```text
starVLA/training/train_starvla.py
approximately lines 324–370
```

In particular, the conditional branch around approximately line 357 appears to skip saving the full model checkpoint when `save_adapter_only` is enabled.

The LoRA adapter is then saved through the VLM-specific adapter-saving path. The relevant implementation appears to be located in:

```text
starVLA/model/modules/vlm/lora_utils.py
approximately around line 393
```

This path appears to call the PEFT saving logic on the module specified by:

```text
qwen_vl_interface.model
```

Since `action_model` is outside this VLM PEFT module, I am unsure whether the updated action-head parameters are preserved through any other saving path when `save_adapter_only: true`.

Could you please clarify the intended behavior?

1. When `save_adapter_only: true`, are updated non-LoRA modules such as `action_model` saved through another path, or is only the VLM LoRA adapter exported?

2. For joint VLM-LoRA and full action-head fine-tuning, should `save_adapter_only` normally remain `false`?

3. When `save_adapter_only: false`, does the full StarVLA checkpoint contain both:

   * the updated VLM LoRA parameters; and
   * the updated `action_model` parameters?

4. When loading the full checkpoint, is the adjacent PEFT adapter directory still required, or is the full checkpoint itself sufficient as long as the LoRA structure is recreated from `config.yaml`?

I may have misunderstood the interaction between the full-checkpoint saving path and the PEFT adapter-saving path, so I would appreciate clarification before starting a long training run.

Thanks.

**Comment by yeahjack** [comment]
@codex review
